src/util/std_expr.cpp

```diff
@@ -223,3 +224,34 @@ const exprt &object_descriptor_exprt::root_object() const
 
   return *p;
 }
+
+void member_exprt::validate(
+  const exprt &expr,
+  const namespacet &ns,
+  const validation_modet vm)
+{
+  check(expr, vm);
+
+  const auto &member_expr = to_member_expr(expr);
+
+  const typet &compound_type = ns.follow(member_expr.compound().type());
+  DATA_CHECK(
+    vm,
+    can_cast_type<struct_union_typet>(compound_type),
+    "member must address a struct, union or compatible type");
+
+  const auto &component = to_struct_union_type(compound_type)
```

Comment:
:question: can't we use a `try_cast_dynamic_type` instead of `can_cast_type` + `to_struct_union_type`?